Allow fall-through in switch

jannotti opened the pull request:
Moved opcode number to be by other "strange jumps"

Added simple tests for normal cases, and for fallthrough

jannotti committed:
Allow fall-through in switch

Moved opcode number to be by other "strange jumps"

Added simple tests for normal cases, and for fallthrough

Added some overflow protection from handcrafted bytecode by limiting
switch labels to 2^16.

michaeldiamant committed:
Replace hard-coded version with AssemblerMaxVersion in assembler_test.go

michaeldiamant committed:
Update docs for switchi fall-through behavior

michaeldiamant committed:
Add switchi assembly test case

michaeldiamant commented on data/transactions/logic/assembler_test.go:
```diff
@@ -410,7 +410,7 @@ const randomnessCompiled = "81ffff03d101d000"
 	randomnessCompiled + "800243218001775c0280018881015d"
 
-const v8Compiled = v7Compiled + pairingCompiled + "8101e002fff800008101"
+const v8Compiled = v7Compiled + pairingCompiled + "81018a02fff800008101"
```
@jannotti Optional - If you prefer, feel welcomed to resolve as is:  I like the convention where we name the nonsense source + compiled stanzas.  I feel like it simplifies reading the construction though I don't feel strongly.

jannotti replied:
I do too, I'll change.  I initially started doing it for things that up in the air about inclusion, so they'd be easy to remove.  But it seems generally better.

michaeldiamant commented on data/transactions/logic/assembler_test.go:
```diff
@@ -410,7 +410,7 @@ const randomnessCompiled = "81ffff03d101d000"
-const v8Compiled = v7Compiled + pairingCompiled + "8101e002fff800008101"
+const v8Compiled = v7Compiled + pairingCompiled + "81018a02fff800008101"
 
 var nonsense = map[uint64]string{
```
@jannotti I have some minor feedback points that I expose via https://github.com/jannotti/go-algorand/pull/12.  Can you have a look?

michaeldiamant replied:
@jannotti Thanks for merging https://github.com/jannotti/go-algorand/pull/12.  I lack resolve permission, but let's consider the thread resolved.

jannotti committed:
Merge pull request #12 from michaeldiamant/switch_fallthrough_pr_review

Feedback on algoidurovic#3 for switchi fall-through

jannotti committed:
Changing encoding to use a single byte for label count

michaeldiamant commented on data/transactions/logic/assembler.go:
```diff
@@ -2565,15 +2567,15 @@ func parseSwitch(program []byte, pos int) (targets []int, nextpc int, err error)
 	pos += bytesUsed
-	if numOffsets > uint64(len(program)) {
-		err = errTooManyItems
+	if numOffsets > math.MaxUint16 {
```
@jannotti Based on https://github.com/algoidurovic/go-algorand/pull/3/commits/c066ff29baec7d5347891e7264388111d15fbd80, I'd expect a limit change here too.  Can you double check me?

* If you take the suggestion:  Note, there's also an import change (remove `math`).  I attached a patch for convenience:  [diff.txt](https://github.com/algoidurovic/go-algorand/files/9578323/diff.txt)
* If you agree with the suggestion, I'd consider applying a constant though I don't feel strongly.


```suggestion
	if numOffsets > 256 {
```

michaeldiamant replied:
https://github.com/algoidurovic/go-algorand/pull/3/commits/d705b0c1d12a53587190a1e763d064d8f647138e resolves the thread for me.  Keep in mind, I don't have resolve access.

algochoi reviewed (commented):
Do we want to bump up the doc version here or later down the road? It looks like the docs weren't regenerated:

https://github.com/jannotti/go-algorand/blob/c066ff29baec7d5347891e7264388111d15fbd80/cmd/opdoc/opdoc.go#L31

```
var docVersion = 7 // -> change to 8
```

jannotti committed:
Fix switch disassembly

Made some changes to cast the label count (numOffsets) to `int`
sooner.  My bug was multiplying numOffsets by 2 while it was still a
byte and _then_ casting to int. It had already wrapped.

michaeldiamant commented:
> Do we want to bump up the doc version here or later down the road? 

@algochoi It's a good callout.  Since AVM 7 is fully out the door, I think we can bump it.  I leave it up to @jannotti - I'd probably do so in an isolated PR because multiple opcodes' docs will be enabled.

jannotti commented:
I think all we need to do first is make sure that devrel (or whoever does it) has taken the v7 docs and put them in place for public consumption.

michaeldiamant commented:
@jannotti https://github.com/algorand/docs/blob/staging/docs/get-details/dapps/avm/teal/specification.md looks updated to me.  As a live proof point, opcodes look updated:  https://developer.algorand.org/docs/get-details/dapps/avm/teal/opcodes/.

The repo recently introduced Github Actions to automatically generate PRs when there's upstream (go-algorand, indexer) releases:  https://github.com/algorand/docs/blob/staging/.github/workflows/release-checker.yml.

Due to the auto PR generation, I think we want to _hold_ on merging updated docs to `master` until ready to release AVM 8.  So, we can either have a long-lived branch/PR previewing doc changes or hold until release gets set in motion.  I don't feel strongly - either option works for me.

jannotti committed:
docVersion and bump pairing ops

jannotti commented:
I bumped the docVersion, moved pairing ops to 9, and put in the v9 nonsense test.
